flink/src/main/java/org/apache/iceberg/flink/actions/MigrateAction.java

```diff
@@ -0,0 +1,423 @@
+
+  private static final String PATQUET_INPUT_FORMAT = "org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat";
+  private static final String ORC_INPUT_FORMAT = "org.apache.hadoop.hive.ql.io.orc.OrcInputFormat";
+  private static final PathFilter HIDDEN_PATH_FILTER =
+      p -> !p.getName().startsWith("_") && !p.getName().startsWith(".");
+  private static final String ICEBERG_METADATA_FOLDER = "metadata";
+
+  private final StreamExecutionEnvironment env;
+  private final HiveCatalog flinkHiveCatalog; // the HiveCatalog of flink
+  private final String hiveSourceDatabaseName;
+  private final String hiveSourceTableName;
+  private final Catalog icebergCatalog;
+  private final Namespace baseNamespace;
+  private final String icebergDatabaseName;
+  private final String icebergTableName;
+  private int maxParallelism;
+
+  public MigrateAction(StreamExecutionEnvironment env, HiveCatalog flinkHiveCatalog, String hiveSourceDatabaseName,
+                       String hiveSourceTableName, Catalog icebergCatalog, Namespace baseNamespace,
+                       String icebergDatabaseName,
+                       String icebergTableName) {
+    this.env = env;
+    this.flinkHiveCatalog = flinkHiveCatalog;
+    this.hiveSourceDatabaseName = hiveSourceDatabaseName;
+    this.hiveSourceTableName = hiveSourceTableName;
+    this.icebergCatalog = icebergCatalog;
+    this.baseNamespace = baseNamespace;
+    this.icebergDatabaseName = icebergDatabaseName;
+    this.icebergTableName = icebergTableName;
+    this.maxParallelism = env.getParallelism();
+  }
+
+  @Override
+  public List<ManifestFile> execute() {
+    flinkHiveCatalog.open();
+    // hive source table
+    ObjectPath tableSource = new ObjectPath(hiveSourceDatabaseName, hiveSourceTableName);
+    org.apache.hadoop.hive.metastore.api.Table hiveTable;
+    try {
+      hiveTable = flinkHiveCatalog.getHiveTable(tableSource);
+    } catch (TableNotExistException e) {
+      throw new RuntimeException(String.format("The source table %s not exists ! ", hiveSourceTableName));
+    }
+
+    List<FieldSchema> fieldSchemaList = hiveTable.getSd().getCols();
+    List<FieldSchema> partitionList = hiveTable.getPartitionKeys();
+    fieldSchemaList.addAll(partitionList);
+    Schema icebergSchema = HiveSchemaUtil.convert(fieldSchemaList);
+    PartitionSpec spec = toPartitionSpec(partitionList, icebergSchema);
+
+    FileFormat fileFormat = getHiveFileFormat(hiveTable);
+
+    TableIdentifier identifier = TableIdentifier.of(toNamespace(), icebergTableName);
+    String hiveLocation = hiveTable.getSd().getLocation();
+
+    Map<String, String> properties = Maps.newHashMap();
+    properties.put(TableProperties.DEFAULT_FILE_FORMAT, fileFormat.name());
+    if (!baseNamespace.isEmpty()) {
+      properties.put(FlinkCatalogFactory.BASE_NAMESPACE, baseNamespace.toString());
+    }
+
+    Table icebergTable;
+    if (icebergCatalog instanceof HadoopCatalog) {
```

Comment:
Why are we using different location for non-HadoopCatalogs?

Reply:
because the `HadoopCatalog`[ has a check](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java#L440), if we set a location for `HadoopCatalog`, it is different from the default,it will throw an exception . so I create the `HadoopCatalog` without  pass the location.

Reply:
In the Spark version of this we just forbid migration to a HadoopCatalog if you attempt to set an alternative location, I think this is basically the same thing. :+1:

Reply:
I think if we add a new Catalog,We should consider whether we should support migration to the new catalog. For example, in the process of doing this migration action, I encountered many problems with the avro format, which was not easy to handle. In addition, I think that avro as a data file format should not be common, so I do not support the avro format .

but I added a verification to make sure `icebergCatalog` is a `hadoopcatalog` or `hivecatalog`.
```
  if (icebergCatalog instanceof HadoopCatalog) {
      icebergTable = icebergCatalog.createTable(identifier, icebergSchema, spec, properties);
    } else if (icebergCatalog instanceof org.apache.iceberg.hive.HiveCatalog) {
      icebergTable = icebergCatalog.createTable(identifier, icebergSchema, spec, hiveLocation, properties);
    }
```

if we add a new Catalog,I think we may need to modify a lot of code，for example : 
https://github.com/apache/iceberg/blob/master/flink/src/test/java/org/apache/iceberg/flink/FlinkCatalogTestBase.java#L104

what do you think ?